// chart-race-react/dist/Bar.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _react = _interopRequireDefault(require("react"));

var _reactTransitionGroup = require("react-transition-group");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(source, true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(source).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

var classes = {
  bar: {
    position: "relative"
  },
  container: {
    width: "100%",
    display: "flex",
    position: "absolute"
  }
};

function Bar(props) {
  var barDefaultStyle = _objectSpread({
    transition: "all ".concat(props.timeout, "ms ease-in-out")
  }, props.prevStyle);

  var posDefaultStyle = {
    transition: "all ".concat(props.timeout, "ms ease-in-out"),
    marginTop: props.prevStyle.marginTop
  };
  var barTransitionStyles = {
    entering: props.prevStyle,
    entered: props.currStyle,
    exiting: props.currStyle
  };
  var posTransitionStyles = {
    entering: {
      marginTop: props.prevStyle.marginTop
    },
    entered: {
      marginTop: props.currStyle.marginTop
    },
    exiting: {
      marginTop: props.currStyle.marginTop
    }
  };
  return _react["default"].createElement("div", {
    style: classes.container
  }, _react["default"].createElement(_reactTransitionGroup.Transition, {
    "in": true,
    timeout: props.timeout
  }, function (state) {
    return _react["default"].createElement(_react["default"].Fragment, null, _react["default"].createElement("div", {
      style: _objectSpread({}, posDefaultStyle, {}, posTransitionStyles[state], {
        width: "".concat(props.width[0], "%")
      })
    }, props.label), _react["default"].createElement("div", {
      style: {
        width: "".concat(props.width[1], "%")
      }
    }, _react["default"].createElement("div", {
      style: _objectSpread({}, classes.bar, {}, barDefaultStyle, {}, barTransitionStyles[state])
    })), _react["default"].createElement("div", {
      style: _objectSpread({}, posDefaultStyle, {}, posTransitionStyles[state], {
        width: "".concat(props.width[2], "%")
      })
    }, _react["default"].createElement("div", {
      style: _objectSpread({}, props.textBoxStyle)
    }, props.value)));
  }));
}

var _default = Bar;
exports["default"] = _default;
